src/ray/raylet/scheduling/local_task_manager_interface.h

```diff
@@ -71,5 +71,65 @@ class ILocalTaskManager {
   virtual size_t GetNumWaitingTaskSpilled() const = 0;
   virtual size_t GetNumUnschedulableTaskSpilled() const = 0;
 };
+
+/// Dummy local task manager, which is used by gcs node only.
+class DummyLocalTaskManager : public ILocalTaskManager {
```

Comment:
I'm not very comfortable with DummyLocalTaskManager here. We should keep nullptr (no perf cost) or just initialize the local task manager with 0 resources and early termination (not sure about the perf cost)

Reply:
Using nullptr for the `local_task_manager_` would leave a lot of `if (local_task_manager_)` branches in the `cluster_task_manager`. This dummy local task manager is light with only two empty hash maps. How do we cost zero resources and what do you mean by early termination?

Reply:
> How do we cost zero resources and what do you mean by the early termination

For example, if the node has resources to run the task, it'll have a worker pool. Can we make the worker pool a raw pointer and if it's NULL, we just early return for some functions? It's also self-contained since if there is no worker pool, it means this node has no resource to execute work, and it'll fail right? There might be other ways.

Reply:
In this case, I feel it might be better to hide the implementation details of local task manager (to cluster task manager), and that is why there is a `ILocalTaskManager` interface here. If using early returns, we have to construct a `LocalTaskManager` (instead of a dummy one), which costs more resources (due to a lot of pointless member variables for a gcs node) and adds more `if..else..` logic inside the `ClusterTaskManager` (or even `SchedulerResourceReporter` and `SchedulerStats` possibly). Thoughts? @wumuzi520 @scv119

Reply:
> which costs more resources


The resources are very little, right?

> adds more if..else.. logic inside the ClusterTaskManager


Compared with Dummy one, it should be less I think?

What I suggested here using the worker pool is just one way and it's not a good way actually.
My overall feeling is that **LocalTaskManager should be able to handle the case where there is no CPU/GPU in the node just for correctness** (I'm happy for any solution with this). Why this will fail? In this way, the scheduler is still coupling with the role (yeah, it's much better than before. To be honest, even with the worker pool solution it's still coupling with the roll, which is why I don't like this either). 

Could you try to make some effort to see whether it's possible to update the LocalTaskManager so that it can handle this case internally without using the constructor or something else? If it's really hard, I'll be ok with a dummy one.

Reply:
At least I think we should rename `DummyLocalTaskManager` to `GcsLocalTaskManager` to make it explicit.

I'm also thinking what should be the correct relationship between cluster_task_manager and local_task_manager. Should cluster_task_manager know local_task_manager at all? Should the upper layer be responsible for moving tasks from cluster_task_manger to local_task_manager? If we do that, we don't have this problem. cc @scv119 @rkooo567 @iycheng

Reply:
>  local_task_manager. Should cluster_task_manager know local_task_manager at all

I think ideally, it shouldn't. But I remember @scv119 mentioned there was a tricky case that we should make cluster task manager aware of local task manager.

Reply:
What's the resolution for this discussion?